modules/partition-replicator/src/main/java/org/apache/ignite/internal/partition/replicator/PartitionReplicaLifecycleManager.java

```diff
@@ -203,22 +225,115 @@ public CompletableFuture<Void> startAsync(ComponentContext componentContext) {
                         inBusyLock(busyLock, () -> onCreateZone(parameters).thenApply((ignored) -> false))
         );
 
-        return nullCompletedFuture();
+        return allOf(processZonesOnStart, processAssignmentsFuture);
+    }
+
+    private CompletableFuture<Void> processZonesOnStart(long recoveryRevision, @Nullable HybridTimestamp lwm) {
+        int earliestCatalogVersion = catalogMgr.activeCatalogVersion(hybridTimestampToLong(lwm));
+        int latestCatalogVersion = catalogMgr.latestCatalogVersion();
+
+        var startedZones = new IntOpenHashSet();
+        var startZoneFutures = new ArrayList<CompletableFuture<?>>();
+
+        for (int ver = latestCatalogVersion; ver >= earliestCatalogVersion; ver--) {
+            int ver0 = ver;
+            catalogMgr.zones(ver).stream()
+                    .filter(tbl -> startedZones.add(tbl.id()))
+                    .forEach(zoneDescriptor -> startZoneFutures.add(
+                            calculateZoneAssignmentsAndCreateReplicationNodes(recoveryRevision, ver0, zoneDescriptor)));
+        }
+
+        return allOf(startZoneFutures.toArray(CompletableFuture[]::new))
+                .whenComplete((unused, throwable) -> {
+                    if (throwable != null) {
+                        LOG.error("Error starting zones", throwable);
+                    } else {
+                        LOG.debug("Zones started successfully");
+                    }
+                });
+    }
+
+    private CompletableFuture<Void> processAssignmentsOnRecovery(long recoveryRevision) {
+        var stableAssignmentsPrefix = new ByteArray(STABLE_ASSIGNMENTS_PREFIX);
+        var pendingAssignmentsPrefix = new ByteArray(PENDING_ASSIGNMENTS_PREFIX);
+
+        CompletableFuture<Void> stableFuture = handleAssignmentsOnRecovery(
```

Comment:
It's tricky why we need to handle stableAssignments on recovery in addition to handling stable within processZonesOnStart. Could you please add a comment here and a common javadoc for processAssignmentsOnRecovery explaining that stable handling here means destroy resources?

Reply:
Comment added `It's required to handle stable assignments changes on recovery in order to cleanup obsolete resources.`